frontend/ui/widget/virtualkeyboard.lua

```diff
@@ -17,9 +17,17 @@ local TextWidget = require("ui/widget/textwidget")
 local UIManager = require("ui/uimanager")
 local VerticalGroup = require("ui/widget/verticalgroup")
 local VerticalSpan = require("ui/widget/verticalspan")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
 local logger = require("logger")
 local Screen = Device.screen
 
+local VirtualKeyPopup = FocusManager:new{
+    modal = true,
+    disable_double_tap = true,
+    inputbox = nil,
+    layout = {},
+}
+
```

Comment:
Why not move all the `VirtualKeyPopup:` methods here ?
I see there's in one of them a reference to `VirtualKey`, so there may be a circular reference that leads to how you laid them. But it looks like it's defined in a function that is created at runtime, so may be it can be looked up fine if you put them here?

Reply:
I could feed the VirtualKeyPopup down in VirtualKeyboard so they're both defined properly.

Reply:
The advantage of VirtualKey:init() is that it has all the modifier etc. logic, so that'd result in some wasted memory on all keys though.

Reply:
Also there's the backspace thing.

Reply:
Maybe as an extra property `hold_callback_default`?

Reply:
Not really following you :)
I just meant: move the code (just the lines, no indentation shifts needed) AFTER the `local VirtualKeyPopup = FocusManager:new{}`, not INSIDE :)
Just to group object and method definition in the file.

Reply:
That can't work because VirtualKey isn't defined yet.

Reply:
I have no idea what you even mean by inside. Where they were before? :-P

Reply:
Fine with me with latest commit :)

Reply:
Alright, merging then so nightly users can test.